### PR TITLE
[DML EP] Add graph capture

### PR DESCRIPTION
**Pull request by PatriceVignola**
This adds a new "Graph Capture" option to the DML ep, similar to the cuda graph functionality. Here's how graph capture works:

- A user can enable graph capture in the session options by setting `ep.dml.enable_graph_capture` to `true`
- When they want to capture a run, they set `gpu_graph_id` in their `RunOptions` to a number bigger than 0 (0 is reserved for internal use according to the cuda graph documentation).
- Then, when they start the inference, the graph will be captured and stored in the DML EP for future use
- When they execute the run for a second time with the same id, the `ReplayGraph` function in the DML EP will be called instead of executing the kernels, resulting in very low overhead and avoiding kernel recompilation.

This feature can give up-to-par or even better performance than specifying the static dimensions at session creation time, but is also much more flexible.

### CODE REVIEW

File: include/onnxruntime/core/providers/dml/dml_provider_factory.h [hunk 1]
```diff
@@ -117,15 +117,16 @@ struct OrtDmlApi {
...
+   * that the supplied command queue is using. The DirectML device and D3D12 command queue must have the same parent ID3D12Device,
+   * or an error will be returned. The D3D12 command queue must be of type DIRECT or COMPUTE (see D3D12_COMMAND_LIST_TYPE). If this
    * function succeeds, the inference session maintains a strong reference on both the dml_device and the command_queue
    * objects.
    * See also: DMLCreateDevice
    * See also: ID3D12Device::CreateCommandQueue
    */
   ORT_API2_STATUS(SessionOptionsAppendExecutionProvider_DML1, _In_ OrtSessionOptions* options,
-                  _In_ IDMLDevice* dml_device, _In_ ID3D12CommandQueue* cmd_queue);
+                  _In_opt_ IDMLDevice* dml_device, _In_ ID3D12CommandQueue* cmd_queue);
```
Comment (sumitsays) [inline include/onnxruntime/core/providers/dml/dml_provider_factory.h:129]: >_In_opt_  [](http://example.com/codeflow?start=18&length=9)

[nit] If sal annotation automatically passes `nullptr` to the optional parameter, then should we keep the optional parameter at the end?

**Review by fdwr — COMMENTED** [review]
\*Some partial comments with the time I have ⌚.

File: onnxruntime/core/providers/dml/DmlExecutionProvider/src/DmlReusedCommandListState.h [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+{
+    struct DmlReusedCommandListState
+    {
+        // Re-usable command list, supporting descriptor heap, and DML binding table to update that heap.
+        Microsoft::WRL::ComPtr<ID3D12GraphicsCommandList> graphicsCommandList;
+        Microsoft::WRL::ComPtr<ID3D12CommandAllocator> commandAllocator;
+        Microsoft::WRL::ComPtr<ID3D12DescriptorHeap> heap;
+        Microsoft::WRL::ComPtr<IDMLBindingTable> bindingTable;
+        Microsoft::WRL::ComPtr<ID3D12Resource> persistentResource;
+        Microsoft::WRL::ComPtr<IUnknown> persistentResourceAllocatorUnk;
```
[the remark this answers is not in the document]
Comment (PatriceVignola) [reply onnxruntime/core/providers/dml/DmlExecutionProvider/src/DmlReusedCommandListState.h:18]: Preexisting :( will fix.

File: include/onnxruntime/core/providers/dml/dml_provider_factory.h [hunk 3]
```diff
@@ -117,8 +117,9 @@ struct OrtDmlApi {
 
   /**
    * Creates a DirectML Execution Provider using the given DirectML device, and which executes work on the supplied D3D12
-   * command queue. The DirectML device and D3D12 command queue must have the same parent ID3D12Device, or an error will
-   * be returned. The D3D12 command queue must be of type DIRECT or COMPUTE (see D3D12_COMMAND_LIST_TYPE). If this
+   * command queue. If a DirectML device isn't provided, a default DirectML device will be created with the same D3D12 device
```
Comment (sumitsays) [inline include/onnxruntime/core/providers/dml/dml_provider_factory.h:120]: >isn't provided, a default DirectML device wil [](http://example.com/codeflow?start=41&length=45)

[nit] Now DML Device will always be provided.

File: onnxruntime/core/providers/dml/dml_provider_factory.cc [hunk 4]
```diff
@@ -31,17 +31,24 @@ using Microsoft::WRL::ComPtr;
...
+  DMLProviderFactory(const ConfigOptions& config_options,
+                     IDMLDevice* dml_device,
                      ID3D12CommandQueue* cmd_queue,
-                     bool disable_metacommands,
-                     bool enable_dynamic_graph_fusion) : dml_device_(dml_device),
+                     bool disable_metacommands) : dml_device_(dml_device),
                                                          cmd_queue_(cmd_queue),
-                                                         metacommands_enabled_(!disable_metacommands),
-                                                         dynamic_graph_fusion_enabled_(enable_dynamic_graph_fusion) {}
+                                                         metacommands_enabled_(!disable_metacommands) {
```
Comment (sumitsays) [inline onnxruntime/core/providers/dml/dml_provider_factory.cc:44]: [nit] Indentation

File: onnxruntime/core/session/inference_session.cc [hunk 5]
```diff
@@ -1748,23 +1749,21 @@ common::Status InferenceSession::Initialize() {
...
 
-        if (dml_graph_fusion_enabled) {
+        if (static_cast<const Dml::ExecutionProvider*>(dmlExecutionProvider)->IsGraphCaptureEnabled()) {
+          std::unique_ptr<onnxruntime::GraphTransformer> dmlRuntimeGraphFusionTransformer = std::make_unique<Dml::DmlRuntimeGraphFusionTransformer>("DmlRuntimeGraphFusionTransformer",
+                                                                                                                                                    dmlExecutionProvider);
+          if (dmlRuntimeGraphFusionTransformer == nullptr) {
+            return Status(common::ONNXRUNTIME, common::FAIL, "DmlRuntimeGraphFusionTransformer is nullptr");
+          }
+          ORT_RETURN_IF_ERROR_SESSIONID_(graph_transformer_mgr_.Register(std::move(dmlRuntimeGraphFusionTransformer), onnxruntime::TransformerLevel::Level3));
+        } else if (dml_graph_fusion_enabled) {
```
Comment (sumitsays) [inline onnxruntime/core/session/inference_session.cc:1759]: Does it mean only one of them can be enabled at a time?

**Review by sumitsays — APPROVED** [review]
:shipit:

File: onnxruntime/core/session/inference_session.cc (hunk 5 above)
Comment (PatriceVignola) [reply onnxruntime/core/session/inference_session.cc:1759]: Yeah, otherwise it partitions them into different partitions and perf ends up being worse.

**Review by sumitsays — APPROVED** [review]
:shipit:

**Review by sumitsays — APPROVED** [review]

File: onnxruntime/core/providers/dml/DmlExecutionProvider/src/ExecutionProvider.cpp [hunk 6]
```diff
@@ -1181,14 +1189,62 @@ namespace Dml
...
     }
 
+    void ExecutionProviderImpl::AppendCapturedGraph(int annotationId, std::unique_ptr<DmlReusedCommandListState> capturedGraph)
+    {
+        m_capturedGraphs[annotationId].push_back(std::move(capturedGraph));
+    }
+
+    onnxruntime::common::Status ExecutionProviderImpl::ReplayGraph(int graph_annotation_id)
+    {
+        for (auto& capturedGraph : m_capturedGraphs[graph_annotation_id])
```
Comment (fdwr) [inline onnxruntime/core/providers/dml/DmlExecutionProvider/src/ExecutionProvider.cpp:1199]:
```suggestion
        for (auto& capturedGraph : m_capturedGraphs[graphAnnotationId])
```
Intra-file self consistency.

And `dml_graph_annotation_id` too below.

File: onnxruntime/core/providers/dml/DmlExecutionProvider/src/ExecutionProvider.cpp [hunk 7]
```diff
@@ -1181,14 +1189,62 @@ namespace Dml
...
+        }
+
+        return onnxruntime::common::Status::OK();
+    }
+
+    Status ExecutionProviderImpl::OnRunStart(const onnxruntime::RunOptions& run_options)
+    {
+        if (GraphCaptureEnabled())
+        {
+            auto graph_annotation_str = run_options.config_options.GetConfigEntry(kOrtRunOptionsConfigCudaGraphAnnotation);
```
Comment (fdwr) [inline onnxruntime/core/providers/dml/DmlExecutionProvider/src/ExecutionProvider.cpp:1211]:
```suggestion
            auto graphAnnotationStr = run_options.config_options.GetConfigEntry(kOrtRunOptionsConfigCudaGraphAnnotation);
```

File: onnxruntime/core/providers/dml/DmlExecutionProvider/src/ExecutionProvider.cpp [hunk 8]
```diff
@@ -1181,14 +1189,62 @@ namespace Dml
...
+            // If graph annotation is not provided, fall back to the one dml graph per session behavior
+            int dml_graph_annotation_id = 0;
```
Comment (fdwr) [inline onnxruntime/core/providers/dml/DmlExecutionProvider/src/ExecutionProvider.cpp:1213]:
```suggestion
            int dmlGraphAnnotationId = 0;
```

**Review by fdwr — DISMISSED** [review]
I don't have the full cognizance of this function to know if any flushes or fences are missing, but it looks okay to me from what I do understand. Just some consistency nits. I didn't look at the Python code.

**Review by fdwr — APPROVED** [review]

**Review by sumitsays — APPROVED** [review]

**Pull request merged**